File: server/python/src/metaobjects/codegen/recover_delegate_emitter.py
"""FR-010 nested-codegen-gap — the runtime-DELEGATING recover emitter (Python).

The self-contained ``recover_<name>(text)`` path (``recover_schema_emitter`` + the
baked ``RecoverSchema``) covers scalars / enums / scalar-arrays but leaves
nested-object and array-of-object components ``None`` — the historical FR-010 codegen
gap. This module emits the additive *delegating* entry that CLOSES that gap by
wrapping the runtime recover:

    recover_<name>_with_loader(root, text, opts=None) -> RecoveryResult[<Name>Recovered]

It resolves this payload's ``MetaObject`` from the supplied loaded ``MetaRoot`` by its
baked simple name (``PAYLOAD_NAME``), delegates to ``recover_object`` in
:mod:`metaobjects.meta.core.object.object_recover` (which assembles the FULL nested
object graph reflection-free via the Phase A object model — ``MetaObject.new_instance()``
+ the ``MetaField`` set-by-name SPI), then maps the assembled ``ValueObject`` graph into
the typed nullable mirror graph via generated ``_from_<vo>_recovered(o)`` mapper
functions (payload + every reachable nested VO, deduped).

This is the codegen-wrapping-runtime pattern (a generated DAO calling the
dynamic-metadata runtime), mirroring the Java / Kotlin / TS pilots. The generated
mappers read the assembled graph through a tiny ``_read_prop`` helper that mirrors the
``MetaField`` get SPI (``ValueObject.get(name)`` else plain-attribute access), so the
emitted code stays self-sufficient and reflection-free.

Bounded by the cross-port ``MAX_NEST_DEPTH`` via the runtime — codegen here only mirrors
the runtime's resolved object graph, so depth/cycle guarding lives in ``object_recover``.
The emitter dedupes mirrors/mappers by VO simple name (cycle-safe).
"""
from __future__ import annotations

from metaobjects.codegen import fr010_field_mapping as fm
from metaobjects.meta.core.field import field_constants as fc
from metaobjects.meta.meta_data import MetaData
from metaobjects.shared.base_types import TYPE_OBJECT
from metaobjects.shared.separators import PACKAGE_SEP


def _find_object(root: MetaData, name: str) -> MetaData | None:
    """The own-child ``object.*`` node named *name*, or ``None``."""
    for c in root.own_children():
        if c.type == TYPE_OBJECT and c.name == name:
            return c
    return None


def _ref_vo(field: MetaData, root: MetaData) -> MetaData | None:
    """The ``@objectRef`` target VO for a nested-object field, or ``None`` when
    unresolvable. Matches first on the full ref, then the trailing simple-name
    segment (mirrors the runtime ``_resolve_object_ref`` short-name fallback)."""
    ref = field.attr(fc.FIELD_ATTR_OBJECT_REF)
    if not isinstance(ref, str) or not ref:
        return None
    direct = _find_object(root, ref)
    if direct is not None:
        return direct
    if PACKAGE_SEP in ref:
        return _find_object(root, ref.rsplit(PACKAGE_SEP, 1)[-1])
    return None


def _is_object_field(field: MetaData) -> bool:
    """True iff the field is a nested object reference (``field.object`` — distinct
    from the string-backed ``field.enum``, treated as a scalar)."""
    return field.sub_type == fc.FIELD_SUBTYPE_OBJECT


def mirror_name(vo: MetaData) -> str:
    """The recovered-mirror dataclass name for a value-object (``<Name>Recovered``)."""
    return f"{vo.name}Recovered"


def _mapper_name(vo: MetaData) -> str:
    """The mapper function name for a value-object (``_from_<name>_recovered``)."""
    return f"_from_{_snake(vo.name)}_recovered"


def root_mapper_name(template_name: str) -> str:
    """The root mapper's name — derived from the TEMPLATE (so it returns the
    canonically-named ``<Template>Recovered`` mirror)."""
    return f"_from_{_snake(template_name)}_recovered"


def _snake(name: str) -> str:
    """``NpcResponseOutput`` → ``npc_response_output`` (matches the cross-generator
    ``_snake_case`` convention; no acronym handling)."""
    out: list[str] = []
    for i, ch in enumerate(name):
        if ch.isupper() and i > 0:
            out.append("_")
        out.append(ch.lower())
    return "".join(out)


# =============================================================================
# Nested-aware mirror type (recurses into nested mirror names)
# =============================================================================


def _nested_mirror_type(field: MetaData, root: MetaData) -> str:
    """The nullable mirror annotation for one field — nested-aware (nested objects
    become ``<Nested>Recovered``; array-of-objects become ``list[...]``)."""
    if _is_object_field(field):
        target = _ref_vo(field, root)
        base = f'"{mirror_name(target)}"' if target is not None else "object"
        elem = f"{base} | None"
        return f"list[{elem}] | None" if fm.is_array(field) else elem
    if fm.is_array(field):
        return "list[str | None] | None"
    if field.sub_type == fc.FIELD_SUBTYPE_ENUM:
        return "str | None"
    kind = fm.scalar_kind(field.sub_type)
    if kind in ("INT", "LONG"):
        return "int | None"
    if kind == "DOUBLE":
        return "float | None"
    if kind == "BOOLEAN":
        return "bool | None"
    return "str | None"


# =============================================================================
# Reachability
# =============================================================================


def reachable_vos(vo: MetaData, root: MetaData) -> list[MetaData]:
    """``vo`` + every value-object reachable through nested ``@objectRef`` fields, in
    stable BFS order, deduped by simple name (cycle-safe)."""
    out: list[MetaData] = []
    seen: set[str] = set()
    queue: list[MetaData] = [vo]
    while queue:
        cur = queue.pop(0)
        if cur.name in seen:
            continue
        seen.add(cur.name)
        out.append(cur)
        for f in fm.fields(cur):
            if _is_object_field(f):
                target = _ref_vo(f, root)
                if target is not None and target.name not in seen:
                    queue.append(target)
    return out


def has_nested(vo: MetaData, root: MetaData) -> bool:
    """True iff ``vo`` (or any reachable nested VO) has a nested-object field."""
    for cur in reachable_vos(vo, root):
        if any(_is_object_field(f) for f in fm.fields(cur)):
            return True
    return False


# =============================================================================
# Nested-aware mirror dataclasses
# =============================================================================


def nested_mirror_dataclasses(
    vo: MetaData, root: MetaData, payload_mirror: str
) -> list[str]:
    """Emit the nested-aware mirror dataclass for ``vo`` and every reachable nested VO
    (deduped). The payload mirror keeps the canonical ``<Template>Recovered`` name
    (``payload_mirror``) so the existing self-contained ``recover_<name>()`` initializer
    and the delegating path share ONE mirror type. The nested mirrors carry their own
    ``<VO>Recovered`` name. Returns source lines (blank-line separated)."""
    lines: list[str] = []
    for i, cur in enumerate(reachable_vos(vo, root)):
        if i > 0:
            lines.append("")
            lines.append("")
        name = payload_mirror if i == 0 else mirror_name(cur)
        lines.extend(_one_mirror(cur, root, name))
    return lines


def _one_mirror(vo: MetaData, root: MetaData, record_name: str) -> list[str]:
    base = (
        record_name[: -len("Recovered")]
        if record_name.endswith("Recovered")
        else record_name
    )
    lines: list[str] = [
        "@dataclass(frozen=True, slots=True)",
        f"class {record_name}:",
        f'    """Best-effort recovered twin of ``{base}`` — every field nullable',
        '    (``None`` where the value was lost or malformed)."""',
    ]
    field_lines = [
        f"    {f.name}: {_nested_mirror_type(f, root)} = None" for f in fm.fields(vo)
    ]
    lines.extend(field_lines or ["    pass"])
    return lines


# =============================================================================
# Mapper functions (assembled ValueObject graph -> typed nullable mirror graph)
# =============================================================================


def nested_mappers(
    vo: MetaData, root: MetaData, root_mapper_fn: str, root_mirror: str
) -> list[str]:
    """Emit one ``_from_<vo>_recovered(o)`` mapper per reachable VO (payload + nested,
    deduped). The ROOT mapper is overridden to the template-derived ``root_mapper_fn`` /
    ``root_mirror`` so it returns the canonically-named root mirror. Returns source
    lines (blank-line separated)."""
    lines: list[str] = []
    vos = reachable_vos(vo, root)
    for i, cur in enumerate(vos):
        if i > 0:
            lines.append("")
            lines.append("")
        fn = root_mapper_fn if i == 0 else _mapper_name(cur)
        mir = root_mirror if i == 0 else mirror_name(cur)
        lines.extend(_one_mapper(cur, root, fn, mir))
    return lines


def _one_mapper(
    vo: MetaData, root: MetaData, fn: str, mirror: str
) -> list[str]:
    lines: list[str] = [
        f'def {fn}(o: object | None) -> "{mirror} | None":',
        f"    \"\"\"Map an assembled ValueObject graph into a typed ``{mirror}`` mirror;"
        " null-tolerant.\"\"\"",
        "    if o is None:",
        "        return None",
        f"    return {mirror}(",
    ]
    for f in fm.fields(vo):
        lines.append(f"        {f.name}={_mapper_arg(f, root)},")
    lines.append("    )")
    return lines


def _mapper_arg(field: MetaData, root: MetaData) -> str:
    """The mirror-field initializer that reads ``field`` from the assembled object ``o``."""
    key = f'"{field.name}"'
    if _is_object_field(field):
        target = _ref_vo(field, root)
        if target is None:
            return "None  # unresolved @objectRef"
        fn = _mapper_name(target)
        if fm.is_array(field):
            return f"_map_object_list(_read_prop(o, {key}), {fn})"
        return f"{fn}(_read_prop(o, {key}))"

    # Enum / scalar / scalar-array: the runtime already coerced; read + light-coerce to
    # the mirror's nullable shape via the locally-defined _dlg_* readers.
    if field.sub_type == fc.FIELD_SUBTYPE_ENUM:
        return f"_dlg_str(_read_prop(o, {key}))"
    if fm.is_array(field):
        return f"_dlg_str_list(_read_prop(o, {key}))"
    kind = fm.scalar_kind(field.sub_type)
    if kind in ("INT", "LONG"):
        return f"_dlg_int(_read_prop(o, {key}))"
    if kind == "DOUBLE":
        return f"_dlg_float(_read_prop(o, {key}))"
    if kind == "BOOLEAN":
        return f"_dlg_bool(_read_prop(o, {key}))"
    return f"_dlg_str(_read_prop(o, {key}))"


# =============================================================================
# Used-helper scoping + the shared helper block
# =============================================================================


def used_helpers(vo: MetaData, root: MetaData) -> set[str]:
    """The generated-helper names the mappers for ``vo`` (+ reachable nested VOs)
    actually reference. ``_read_prop`` is always present once any mapper is emitted."""
    used: set[str] = {"_read_prop"}
    for cur in reachable_vos(vo, root):
        for f in fm.fields(cur):
            if _is_object_field(f):
                if fm.is_array(f):
                    used.add("_map_object_list")
                continue
            if f.sub_type == fc.FIELD_SUBTYPE_ENUM:
                used.add("_dlg_str")
            elif fm.is_array(f):
                used.add("_dlg_str_list")
            else:
                kind = fm.scalar_kind(f.sub_type)
                if kind in ("INT", "LONG"):
                    used.add("_dlg_int")
                elif kind == "DOUBLE":
                    used.add("_dlg_float")
                elif kind == "BOOLEAN":
                    used.add("_dlg_bool")
                else:
                    used.add("_dlg_str")
    return used


# Each helper: name -> source block (lines).
_HELPER_BLOCKS: dict[str, list[str]] = {
    "_read_prop": [
        "def _read_prop(o: object | None, name: str) -> object | None:",
        '    """Read a property from an assembled backing object, mirroring the',
        "    MetaField get SPI (``ValueObject.get(name)`` else plain-attribute access).",
        '    Keeps the mappers reflection-free + backing-agnostic."""',
        "    if o is None:",
        "        return None",
        '    getter = getattr(o, "get", None)',
        "    if callable(getter):",
        "        return getter(name)",
        "    return getattr(o, name, None)",
    ],
    "_map_object_list": [
        "def _map_object_list(v: object | None, fn) -> list | None:",
        '    """Map each element of an assembled array via ``fn``; non-list -> None."""',
        "    if not isinstance(v, list):",
        "        return None",
        "    return [fn(e) for e in v]",
    ],
    "_dlg_str": [
        "def _dlg_str(v: object | None) -> str | None:",
        "    return None if v is None else str(v)",
    ],
    "_dlg_int": [
        "def _dlg_int(v: object | None) -> int | None:",
        "    if v is None:",
        "        return None",
        "    try:",
        "        return int(v)  # type: ignore[arg-type, call-overload]",
        "    except (TypeError, ValueError):",
        "        return None",
    ],
    "_dlg_float": [
        "def _dlg_float(v: object | None) -> float | None:",
        "    if v is None:",
        "        return None",
        "    try:",
        "        return float(v)  # type: ignore[arg-type]",
        "    except (TypeError, ValueError):",
        "        return None",
    ],
    "_dlg_bool": [
        "def _dlg_bool(v: object | None) -> bool | None:",
        "    if v is None:",
        "        return None",
        "    if isinstance(v, bool):",
        "        return v",
        '    return str(v).lower() == "true"',
    ],
    "_dlg_str_list": [
        "def _dlg_str_list(v: object | None) -> list | None:",
        "    if not isinstance(v, list):",
        "        return None",
        "    return [None if e is None else str(e) for e in v]",
    ],
}

# Stable emission order (deterministic output).
_HELPER_ORDER: tuple[str, ...] = (
    "_read_prop",
    "_map_object_list",
    "_dlg_str",
    "_dlg_int",
    "_dlg_float",
    "_dlg_bool",
    "_dlg_str_list",
)


def delegate_helpers(used: set[str]) -> list[str]:
    """The shared helper block the generated mappers rely on, scoped to ``used`` and
    emitted in stable order. Returns source lines (blank-line separated)."""
    lines: list[str] = [
        "# ---- runtime-delegating recover helpers (generated) ----"
    ]
    for name in _HELPER_ORDER:
        if name in used:
            lines.append("")
            lines.append("")
            lines.extend(_HELPER_BLOCKS[name])
    return lines

File: server/csharp/MetaObjects.Codegen/Generators/RecoverDelegateEmitter.cs
// FR-010 Plan 2.1 (nested codegen gap) — the runtime-DELEGATING recover emitter (C# port).
//
// The self-contained Recover(string) path (RecoverSchemaEmitter + the baked RecoverSchema +
// RecoverMap reads) covers scalars / enums / scalar-arrays but leaves nested-object and
// array-of-object components NULL — the historical FR-010 codegen gap. This module emits the
// additive runtime-DELEGATING overload that CLOSES that gap by wrapping the runtime recover:
//
//   Recover(MetaObject mo, string text[, RecoverOptions opts]) -> RecoveryResult<<Name>Recovered>
//
// The delegating overload takes a runtime MetaObject (resolved by the caller from a
// MetaDataLoader's Root via its baked PAYLOAD_FQN — a convenience MetaDataLoader overload is
// emitted too) and delegates to MetaObjects.Codegen.Runtime.RecoverObject.Recover(mo, text,
// Format, opts), which assembles the FULL nested object graph reflection-free via the Phase A
// object model (MetaObject.NewInstance() + the MetaField SPI). It then maps the assembled
// ValueObject graph into the typed nullable mirror graph via generated From<VO>Recovered(...)
// mapper methods (payload + every reachable nested VO, deduped).
//
// This is the codegen-wrapping-runtime pattern (a generated DAO calling the dynamic-metadata
// runtime), mirroring the Java SpringOutputParserGenerator + Kotlin/TS/Python pilots. The
// generated mappers read the assembled graph through a tiny ReadProp() helper mirroring the
// MetaField GetValue SPI (ValueObject.Get(name)), so the emitted code stays self-sufficient and
// reflection-free / AOT-safe.

using System.Collections.Generic;
using System.Linq;
using System.Text;
using MetaObjects.Meta;
using static MetaObjects.Shared.BaseTypes;
using static MetaObjects.Shared.Structural;
using static MetaObjects.Core.Field.FieldConstants;

namespace MetaObjects.Codegen.Generators;

internal static class RecoverDelegateEmitter
{
    // =========================================================================
    // VO / field discovery (object-before-isArray order — the cross-port fix)
    // =========================================================================

    private static MetaData? FindObject(MetaData root, string name) =>
        root.OwnChildren().FirstOrDefault(c => c.Type == TYPE_OBJECT && c.Name == name);

    /// <summary>
    /// The <c>@objectRef</c> target VO for a nested-object field, or null when unresolvable.
    /// Matches by the bare ref first, then by the package-stripped short name.
    /// </summary>
    private static MetaData? RefVo(MetaData field, MetaData root)
    {
        if (field.OwnAttr(FIELD_ATTR_OBJECT_REF) is not string objectRef) return null;
        var direct = FindObject(root, objectRef);
        if (direct is not null) return direct;
        int sep = objectRef.LastIndexOf(PACKAGE_SEPARATOR, System.StringComparison.Ordinal);
        return sep >= 0 ? FindObject(root, objectRef[(sep + PACKAGE_SEPARATOR.Length)..]) : null;
    }

    /// <summary>
    /// True iff the field is a nested object reference (<c>field.object</c>) — distinct from the
    /// string-backed <c>field.enum</c>, which is treated as a scalar. OBJECT is tested BEFORE
    /// IsArray (the object-before-isArray order) so an array-of-objects maps to nested mirrors,
    /// not a string list.
    /// </summary>
    private static bool IsObjectField(MetaData field) => field.SubType == FIELD_SUBTYPE_OBJECT;

    /// <summary>The recovered-mirror record name for a value-object (<c>&lt;Name&gt;Recovered</c>).</summary>
    public static string MirrorName(MetaData vo) => $"{vo.Name}Recovered";

    /// <summary>The mapper-method name for a value-object (<c>From&lt;Name&gt;Recovered</c>).</summary>
    private static string MapperName(MetaData vo) => $"From{vo.Name}Recovered";

    // =========================================================================
    // "Has nested" — only emit the delegating overload + mappers when worthwhile
    // =========================================================================

    /// <summary>True iff the payload (or any reachable nested VO) has a nested-object / array-of-object field.</summary>
    public static bool HasNested(MetaData vo, MetaData root)
    {
        var seen = new HashSet<string>(System.StringComparer.Ordinal);
        var stack = new Stack<MetaData>();
        stack.Push(vo);
        while (stack.Count > 0)
        {
            var cur = stack.Pop();
            if (!seen.Add(cur.Name)) continue;
            foreach (var f in Fr010FieldMapping.Fields(cur))
            {
                if (!IsObjectField(f)) continue;
                var target = RefVo(f, root);
                if (target is not null) stack.Push(target);
                return true; // a nested object field exists regardless of resolvability
            }
        }
        return false;
    }

    // =========================================================================
    // Nested-aware mirror records (payload + every reachable nested VO)
    // =========================================================================

    /// <summary>The nullable mirror C# type for one field — nested-aware (recurses into nested mirror names).</summary>
    private static string NestedMirrorType(MetaData field, MetaData root)
    {
        // Object BEFORE array (the object-before-isArray fix): an array-of-objects must map to
        // a list of nested mirrors, NOT a string list.
        if (IsObjectField(field))
        {
            var target = RefVo(field, root);
            string baseName = target is not null ? MirrorName(target) : "object";
            return Fr010FieldMapping.IsArray(field)
                ? $"global::System.Collections.Generic.IReadOnlyList<{baseName}?>?"
                : $"{baseName}?";
        }
        if (Fr010FieldMapping.IsArray(field))
            return "global::System.Collections.Generic.IReadOnlyList<string?>?";
        if (field.SubType == FIELD_SUBTYPE_ENUM) return "string?";
        return Fr010FieldMapping.ScalarKind(field.SubType) switch
        {
            "Int" => "int?",
            "Long" => "long?",
            "Double" => "double?",
            "Boolean" => "bool?",
            _ => "string?",
        };
    }

    /// <summary>
    /// Emit the nested-aware mirror record for <paramref name="vo"/> and every value-object
    /// reachable from it (deduped by simple name; cycle-safe). The PAYLOAD mirror keeps the
    /// canonical <c>&lt;Payload&gt;Recovered</c> name (<paramref name="payloadMirror"/>) so the
    /// self-contained <c>Recover(string)</c> and the delegating overload share one mirror type.
    /// The nested mirror records are emitted here (the self-contained emitter only emits the
    /// payload mirror).
    /// </summary>
    /// <summary>
    /// Emit the PAYLOAD mirror record (nested-aware, so object fields are typed as nested mirrors,
    /// not <c>object?</c>) plus every reachable NESTED mirror record. Replaces the self-contained
    /// <see cref="RecoverSchemaEmitter.MirrorRecordDecl"/> on the delegating path so the one shared
    /// <c>&lt;Payload&gt;Recovered</c> type can carry populated nested components.
    /// </summary>
    public static string NestedMirrorRecords(MetaData vo, MetaData root, string payloadMirror)
    {
        var sb = new StringBuilder();
        var seen = new HashSet<string>(System.StringComparer.Ordinal);
        EmitMirror(vo, root, payloadMirror, seen, sb);
        return sb.ToString();
    }

    private static void EmitMirror(MetaData vo, MetaData root, string recordName,
        HashSet<string> seen, StringBuilder sb)
    {
        if (!seen.Add(vo.Name)) return;
        string baseName = recordName.EndsWith("Recovered", System.StringComparison.Ordinal)
            ? recordName[..^"Recovered".Length] : recordName;
        sb.AppendLine();
        sb.AppendLine($"/// <summary>Best-effort recovered twin of <c>{baseName}</c> — every component nullable (null where lost/malformed).</summary>");
        sb.AppendLine($"public sealed record {recordName}");
        sb.AppendLine("{");
        foreach (var f in Fr010FieldMapping.Fields(vo))
            sb.AppendLine($"    public {NestedMirrorType(f, root)} {f.Name} {{ get; init; }}");
        sb.AppendLine("}");

        foreach (var f in Fr010FieldMapping.Fields(vo))
            if (IsObjectField(f) && RefVo(f, root) is { } target)
                EmitMirror(target, root, MirrorName(target), seen, sb);
    }

    // =========================================================================
    // The delegating recover overload + ValueObject(map)->mirror mappers
    // =========================================================================

    /// <summary>
    /// Emit the runtime-delegating recover members appended inside the generated parser class:
    /// the <c>PAYLOAD_FQN</c> constant, the <c>Recover(MetaObject, ...)</c> + convenience
    /// <c>Recover(MetaDataLoader, ...)</c> overloads, the recursive <c>From&lt;VO&gt;Recovered</c>
    /// mappers (payload + nested, deduped), and the shared <c>ReadProp</c> / <c>MapObjectList</c>
    /// / <c>Dlg*</c> helpers.
    /// </summary>
    public static string DelegatingMembers(MetaData vo, MetaData root, string payloadFqn,
        string rootMirror, string formatEnum)
    {
        var sb = new StringBuilder();
        sb.AppendLine();
        sb.AppendLine("    // FR-010 Plan 2.1 — runtime-delegating recovery (closes the nested-object / array-of-object");
        sb.AppendLine("    // gap left by the self-contained Recover(string) path). Delegates to the runtime");
        sb.AppendLine("    // MetaObjects.Codegen.Runtime.RecoverObject, which assembles the FULL nested object graph");
        sb.AppendLine("    // reflection-free, then maps it into the typed nullable mirror via From*Recovered.");
        sb.AppendLine();
        sb.AppendLine($"    /// <summary>The payload's metadata name — resolve it against a loaded <c>MetaRoot</c> to obtain the runtime <c>MetaObject</c>.</summary>");
        sb.AppendLine($"    public const string PAYLOAD_FQN = \"{Fr010FieldMapping.CSharpStringLiteral(payloadFqn)}\";");
        sb.AppendLine();
        sb.AppendLine($"    /// <summary>Tolerant best-effort recovery delegating to the runtime; fully populates nested-object and");
        sb.AppendLine($"    /// array-of-object components (unlike the self-contained <see cref=\"Recover(string)\"/>). Never throws.</summary>");
        sb.AppendLine($"    public static global::MetaObjects.Render.Recover.RecoveryResult<{rootMirror}> Recover(");
        sb.AppendLine($"        global::MetaObjects.Meta.MetaObject mo, string text, RecoverOptions? opts = null)");
        sb.AppendLine("    {");
        sb.AppendLine($"        var raw = global::MetaObjects.Codegen.Runtime.RecoverObject.Recover(mo, text, {formatEnum}, opts);");
        sb.AppendLine("        // The assembled graph is a ValueObject (nested ValueObjects / lists of them) — map it into");
        sb.AppendLine("        // the typed mirror graph. ReadProp mirrors the MetaField GetValue SPI (reflection-free).");
        sb.AppendLine($"        var data = {vo.Name}RecoveredRoot(raw.Data);");
        sb.AppendLine($"        return new global::MetaObjects.Render.Recover.RecoveryResult<{rootMirror}>(data, raw.Report);");
        sb.AppendLine("    }");
        sb.AppendLine();
        sb.AppendLine($"    /// <summary>Convenience overload — resolves <see cref=\"PAYLOAD_FQN\"/> from a loaded <c>MetaRoot</c>, then delegates.</summary>");
        sb.AppendLine($"    public static global::MetaObjects.Render.Recover.RecoveryResult<{rootMirror}> Recover(");
        sb.AppendLine($"        global::MetaObjects.Meta.MetaRoot root, string text, RecoverOptions? opts = null)");
        sb.AppendLine("    {");
        sb.AppendLine("        var mo = root.FindObject(PAYLOAD_FQN)");
        sb.AppendLine("            ?? throw new global::System.InvalidOperationException(");
        sb.AppendLine("                $\"payload object \\\"{PAYLOAD_FQN}\\\" not found in the loaded metadata\");");
        sb.AppendLine("        return Recover(mo, text, opts);");
        sb.AppendLine("    }");

        // ---- mappers (root + nested, deduped). Root mapper is named distinctly so it can carry
        //      the canonical payload-mirror return type (the template name may differ from the VO).
        EmitMappers(sb, vo, root, rootMirror);

        // ---- shared helpers
        AppendHelpers(sb);
        return sb.ToString();
    }

    private static void EmitMappers(StringBuilder sb, MetaData rootVo, MetaData root, string rootMirror)
    {
        var seen = new HashSet<string>(System.StringComparer.Ordinal);
        // Root mapper: a distinctly-named method returning the canonical payload mirror.
        EmitRootMapper(sb, rootVo, root, rootMirror);
        seen.Add(rootVo.Name);
        // Nested mappers (each named From<VO>Recovered, returning <VO>Recovered).
        foreach (var f in Fr010FieldMapping.Fields(rootVo))
            if (IsObjectField(f) && RefVo(f, root) is { } target)
                EmitNestedMapper(sb, target, root, seen);
    }

    private static void EmitRootMapper(StringBuilder sb, MetaData vo, MetaData root, string mirror)
    {
        sb.AppendLine();
        sb.AppendLine($"    /// <summary>Map an assembled ValueObject graph into a typed <c>{mirror}</c>. Generated; null-tolerant.</summary>");
        sb.AppendLine($"    private static {mirror}? {vo.Name}RecoveredRoot(object? o)");
        sb.AppendLine("    {");
        sb.AppendLine("        if (o is null) return null;");
        sb.AppendLine($"        return new {mirror}");
        sb.AppendLine("        {");
        foreach (var f in Fr010FieldMapping.Fields(vo))
            sb.AppendLine($"            {f.Name} = {MapperArg(f, root)},");
        sb.AppendLine("        };");
        sb.AppendLine("    }");
    }

    private static void EmitNestedMapper(StringBuilder sb, MetaData vo, MetaData root, HashSet<string> seen)
    {
        if (!seen.Add(vo.Name)) return;
        string mirror = MirrorName(vo);
        sb.AppendLine();
        sb.AppendLine($"    /// <summary>Map an assembled ValueObject graph into a typed <c>{mirror}</c>. Generated; null-tolerant.</summary>");
        sb.AppendLine($"    private static {mirror}? {MapperName(vo)}(object? o)");
        sb.AppendLine("    {");
        sb.AppendLine("        if (o is null) return null;");
        sb.AppendLine($"        return new {mirror}");
        sb.AppendLine("        {");
        foreach (var f in Fr010FieldMapping.Fields(vo))
            sb.AppendLine($"            {f.Name} = {MapperArg(f, root)},");
        sb.AppendLine("        };");
        sb.AppendLine("    }");

        foreach (var f in Fr010FieldMapping.Fields(vo))
            if (IsObjectField(f) && RefVo(f, root) is { } target)
                EmitNestedMapper(sb, target, root, seen);
    }

    /// <summary>The mirror-field initializer expression that reads <paramref name="field"/> from the assembled object.</summary>
    private static string MapperArg(MetaData field, MetaData root)
    {
        string key = $"\"{Fr010FieldMapping.CSharpStringLiteral(field.Name)}\"";

        // Object BEFORE array (object-before-isArray): array-of-objects maps to nested mirrors.
        if (IsObjectField(field))
        {
            var target = RefVo(field, root);
            if (target is null) return "null /* unresolved @objectRef */";
            string fn = MapperName(target);
            return Fr010FieldMapping.IsArray(field)
                ? $"MapObjectList(ReadProp(o, {key}), {fn})"
                : $"{fn}(ReadProp(o, {key}))";
        }

        // Enum / scalar / scalar-array: the runtime already coerced; read + light-coerce via Dlg*.
        if (field.SubType == FIELD_SUBTYPE_ENUM) return $"DlgString(ReadProp(o, {key}))";
        if (Fr010FieldMapping.IsArray(field)) return $"DlgStringList(ReadProp(o, {key}))";
        return Fr010FieldMapping.ScalarKind(field.SubType) switch
        {
            "Int" => $"DlgInt(ReadProp(o, {key}))",
            "Long" => $"DlgLong(ReadProp(o, {key}))",
            "Double" => $"DlgDouble(ReadProp(o, {key}))",
            "Boolean" => $"DlgBool(ReadProp(o, {key}))",
            _ => $"DlgString(ReadProp(o, {key}))",
        };
    }

    /// <summary>
    /// Append the shared runtime-delegating helpers the mappers rely on. <c>ReadProp</c> mirrors
    /// the MetaField GetValue SPI (<c>ValueObject.Get(name)</c>) so the mappers stay reflection-free
    /// + backing-agnostic; <c>MapObjectList</c> maps an assembled list element-wise; the <c>Dlg*</c>
    /// readers light-coerce to the mirror's nullable scalar shapes. (The <c>Dlg</c> prefix avoids
    /// shadowing the render <c>RecoverMap.As*(d, key)</c> two-arg readers the self-contained path uses.)
    /// </summary>
    private static void AppendHelpers(StringBuilder sb)
    {
        sb.AppendLine();
        sb.AppendLine("    // ---- runtime-delegating recover helpers (generated) ----");
        sb.AppendLine();
        sb.AppendLine("    /// <summary>Read a property from an assembled backing object, mirroring the MetaField GetValue SPI.</summary>");
        sb.AppendLine("    private static object? ReadProp(object? o, string name) =>");
        sb.AppendLine("        o is global::MetaObjects.Meta.ValueObject vo ? vo.Get(name) : null;");
        sb.AppendLine();
        sb.AppendLine("    /// <summary>Map each element of an assembled list via <paramref name=\"fn\"/>; null/absent -> null.</summary>");
        sb.AppendLine("    private static global::System.Collections.Generic.IReadOnlyList<T?>? MapObjectList<T>(");
        sb.AppendLine("        object? v, global::System.Func<object?, T?> fn) where T : class");
        sb.AppendLine("    {");
        sb.AppendLine("        if (v is not global::System.Collections.IEnumerable e || v is string) return null;");
        sb.AppendLine("        var outList = new global::System.Collections.Generic.List<T?>();");
        sb.AppendLine("        foreach (var elem in e) outList.Add(fn(elem));");
        sb.AppendLine("        return outList;");
        sb.AppendLine("    }");
        sb.AppendLine();
        sb.AppendLine("    private static string? DlgString(object? v) => v?.ToString();");
        sb.AppendLine();
        sb.AppendLine("    private static int? DlgInt(object? v) => v switch");
        sb.AppendLine("    {");
        sb.AppendLine("        null => null,");
        sb.AppendLine("        int i => i,");
        sb.AppendLine("        long l => unchecked((int)l),");
        sb.AppendLine("        global::System.IConvertible => global::System.Convert.ToInt32(v, global::System.Globalization.CultureInfo.InvariantCulture),");
        sb.AppendLine("        _ => int.TryParse(v.ToString(), out var n) ? n : (int?)null,");
        sb.AppendLine("    };");
        sb.AppendLine();
        sb.AppendLine("    private static long? DlgLong(object? v) => v switch");
        sb.AppendLine("    {");
        sb.AppendLine("        null => null,");
        sb.AppendLine("        long l => l,");
        sb.AppendLine("        int i => i,");
        sb.AppendLine("        global::System.IConvertible => global::System.Convert.ToInt64(v, global::System.Globalization.CultureInfo.InvariantCulture),");
        sb.AppendLine("        _ => long.TryParse(v.ToString(), out var n) ? n : (long?)null,");
        sb.AppendLine("    };");
        sb.AppendLine();
        sb.AppendLine("    private static double? DlgDouble(object? v) => v switch");
        sb.AppendLine("    {");
        sb.AppendLine("        null => null,");
        sb.AppendLine("        double d => d,");
        sb.AppendLine("        global::System.IConvertible => global::System.Convert.ToDouble(v, global::System.Globalization.CultureInfo.InvariantCulture),");
        sb.AppendLine("        _ => double.TryParse(v.ToString(), global::System.Globalization.NumberStyles.Any, global::System.Globalization.CultureInfo.InvariantCulture, out var n) ? n : (double?)null,");
        sb.AppendLine("    };");
        sb.AppendLine();
        sb.AppendLine("    private static bool? DlgBool(object? v) => v switch");
        sb.AppendLine("    {");
        sb.AppendLine("        null => null,");
        sb.AppendLine("        bool b => b,");
        sb.AppendLine("        _ => string.Equals(v.ToString(), \"true\", global::System.StringComparison.OrdinalIgnoreCase),");
        sb.AppendLine("    };");
        sb.AppendLine();
        sb.AppendLine("    /// <summary>Map an assembled list of scalars into a nullable-element string list.</summary>");
        sb.AppendLine("    private static global::System.Collections.Generic.IReadOnlyList<string?>? DlgStringList(object? v)");
        sb.AppendLine("    {");
        sb.AppendLine("        if (v is not global::System.Collections.IEnumerable e || v is string) return null;");
        sb.AppendLine("        var outList = new global::System.Collections.Generic.List<string?>();");
        sb.AppendLine("        foreach (var elem in e) outList.Add(elem?.ToString());");
        sb.AppendLine("        return outList;");
        sb.AppendLine("    }");
    }
}
